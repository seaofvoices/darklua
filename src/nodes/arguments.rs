use std::iter;

use crate::nodes::{Expression, StringExpression, TableExpression, Token};

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct TupleArgumentsTokens {
    pub opening_parenthese: Token,
    pub closing_parenthese: Token,
    pub commas: Vec<Token>,
}

impl TupleArgumentsTokens {
    super::impl_token_fns!(
        target = [opening_parenthese, closing_parenthese]
        iter = [commas]
    );
}

#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct TupleArguments {
    values: Vec<Expression>,
    tokens: Option<TupleArgumentsTokens>,
}

impl TupleArguments {
    pub fn new(values: Vec<Expression>) -> Self {
        Self {
            values,
            tokens: None,
        }
    }

    #[inline]
    pub fn to_expressions(self) -> Vec<Expression> {
        self.values
    }

    pub fn with_tokens(mut self, tokens: TupleArgumentsTokens) -> Self {
        self.tokens = Some(tokens);
        self
    }

    #[inline]
    pub fn set_tokens(&mut self, tokens: TupleArgumentsTokens) {
        self.tokens = Some(tokens);
    }

    #[inline]
    pub fn get_tokens(&self) -> Option<&TupleArgumentsTokens> {
        self.tokens.as_ref()
    }

    pub fn with_argument<T: Into<Expression>>(mut self, argument: T) -> Self {
        self.values.push(argument.into());
        self
    }

    #[inline]
    pub fn len(&self) -> usize {
        self.values.len()
    }

    #[inline]
    pub fn is_empty(&self) -> bool {
        self.values.is_empty()
    }

    #[inline]
    pub fn iter_values(&self) -> impl Iterator<Item = &Expression> {
        self.values.iter()
    }

    #[inline]
    pub fn iter_mut_values(&mut self) -> impl Iterator<Item = &mut Expression> {
        self.values.iter_mut()
    }

    super::impl_token_fns!(iter = [tokens]);
}

impl From<Arguments> for TupleArguments {
    fn from(arguments: Arguments) -> Self {
        match arguments {
            Arguments::Tuple(tuple) => tuple,
            Arguments::String(string) => TupleArguments::default().with_argument(string),
            Arguments::Table(table) => TupleArguments::default().with_argument(table),
        }
    }
}

impl iter::FromIterator<Expression> for TupleArguments {
    fn from_iter<T: IntoIterator<Item = Expression>>(iter: T) -> Self {
        Self {
            values: iter.into_iter().collect(),
            tokens: None,
        }
    }
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum Arguments {
    Tuple(TupleArguments),
    String(StringExpression),
    Table(TableExpression),
}

impl Arguments {
    pub fn to_expressions(self) -> Vec<Expression> {
        match self {
            Self::Tuple(expressions) => expressions.to_expressions(),
            Self::String(string) => vec![string.into()],
            Self::Table(table) => vec![table.into()],
        }
    }

    pub fn with_argument<T: Into<Expression>>(self, argument: T) -> Self {
        TupleArguments::from(self).with_argument(argument).into()
    }

    pub fn clear_comments(&mut self) {
        match self {
            Arguments::Tuple(tuple) => tuple.clear_comments(),
            Arguments::String(_) | Arguments::Table(_) => {}
        }
    }

    pub fn clear_whitespaces(&mut self) {
        match self {
            Arguments::Tuple(tuple) => tuple.clear_whitespaces(),
            Arguments::String(_) | Arguments::Table(_) => {}
        }
    }

    pub(crate) fn replace_referenced_tokens(&mut self, code: &str) {
        match self {
            Arguments::Tuple(tuple) => tuple.replace_referenced_tokens(code),
            Arguments::String(_) | Arguments::Table(_) => {}
        }
    }

    pub(crate) fn shift_token_line(&mut self, amount: isize) {
        match self {
            Arguments::Tuple(tuple) => tuple.shift_token_line(amount),
            Arguments::String(_) | Arguments::Table(_) => {}
        }
    }

    pub(crate) fn filter_comments(&mut self, filter: impl Fn(&super::Trivia) -> bool) {
        match self {
            Arguments::Tuple(tuple) => tuple.filter_comments(filter),
            Arguments::String(_) | Arguments::Table(_) => {}
        }
    }
}

impl Default for Arguments {
    fn default() -> Self {
        Self::Tuple(TupleArguments::default())
    }
}

impl From<TupleArguments> for Arguments {
    fn from(tuple: TupleArguments) -> Self {
        Self::Tuple(tuple)
    }
}

impl From<TableExpression> for Arguments {
    fn from(table: TableExpression) -> Self {
        Self::Table(table)
    }
}

impl From<StringExpression> for Arguments {
    fn from(string: StringExpression) -> Self {
        Self::String(string)
    }
}
