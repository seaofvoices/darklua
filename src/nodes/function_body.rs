use super::{
    Block, FunctionExpression, FunctionName, FunctionReturnType, FunctionStatement,
    FunctionVariadicType, GenericParameters, Identifier, LocalFunctionStatement,
    LocalFunctionTokens, Token, TypedIdentifier,
};

pub(crate) struct FunctionBuilder {
    block: Block,
    parameters: Vec<TypedIdentifier>,
    is_variadic: bool,
    variadic_type: Option<FunctionVariadicType>,
    return_type: Option<FunctionReturnType>,

    function: Option<Token>,
    opening_parenthese: Option<Token>,
    closing_parenthese: Option<Token>,
    end: Option<Token>,
    parameter_commas: Vec<Token>,

    variable_arguments: Option<Token>,
    variable_arguments_colon: Option<Token>,
    return_type_colon: Option<Token>,
    generic_parameters: Option<GenericParameters>,
}

impl FunctionBuilder {
    pub fn from_block(block: impl Into<Block>) -> Self {
        Self {
            block: block.into(),
            parameters: Vec::new(),
            is_variadic: false,
            variadic_type: None,
            return_type: None,

            function: None,
            opening_parenthese: None,
            closing_parenthese: None,
            end: None,
            parameter_commas: Vec::new(),

            variable_arguments: None,
            variable_arguments_colon: None,
            return_type_colon: None,
            generic_parameters: None,
        }
    }

    pub(crate) fn into_function_expression(self) -> FunctionExpression {
        let mut expression = FunctionExpression::new(self.block, self.parameters, self.is_variadic);

        if let Some(variadic_type) = self.variadic_type {
            expression.set_variadic_type(variadic_type);
        }

        if let Some(return_type) = self.return_type {
            expression.set_return_type(return_type);
        }

        if let Some(generic_parameters) = self.generic_parameters {
            expression.set_generic_parameters(generic_parameters);
        }

        if let (Some(function), Some(opening_parenthese), Some(closing_parenthese), Some(end)) = (
            self.function,
            self.opening_parenthese,
            self.closing_parenthese,
            self.end,
        ) {
            expression.set_tokens(FunctionBodyTokens {
                function,
                opening_parenthese,
                closing_parenthese,
                end,
                parameter_commas: self.parameter_commas,
                variable_arguments: self.variable_arguments,
                variable_arguments_colon: self.variable_arguments_colon,
                return_type_colon: self.return_type_colon,
            });
        }

        expression
    }

    pub(crate) fn into_function_statement(self, name: FunctionName) -> FunctionStatement {
        let mut statement =
            FunctionStatement::new(name, self.block, self.parameters, self.is_variadic);

        if let Some(variadic_type) = self.variadic_type {
            statement.set_variadic_type(variadic_type);
        }

        if let Some(return_type) = self.return_type {
            statement.set_return_type(return_type);
        }

        if let Some(generic_parameters) = self.generic_parameters {
            statement.set_generic_parameters(generic_parameters);
        }

        if let (Some(function), Some(opening_parenthese), Some(closing_parenthese), Some(end)) = (
            self.function,
            self.opening_parenthese,
            self.closing_parenthese,
            self.end,
        ) {
            statement.set_tokens(FunctionBodyTokens {
                function,
                opening_parenthese,
                closing_parenthese,
                end,
                parameter_commas: self.parameter_commas,
                variable_arguments: self.variable_arguments,
                variable_arguments_colon: self.variable_arguments_colon,
                return_type_colon: self.return_type_colon,
            });
        }

        statement
    }

    pub(crate) fn into_local_function_statement(
        self,
        name: Identifier,
        local_token: Option<Token>,
    ) -> LocalFunctionStatement {
        let mut statement =
            LocalFunctionStatement::new(name, self.block, self.parameters, self.is_variadic);

        if let Some(variadic_type) = self.variadic_type {
            statement.set_variadic_type(variadic_type);
        }

        if let Some(return_type) = self.return_type {
            statement.set_return_type(return_type);
        }

        if let Some(generic_parameters) = self.generic_parameters {
            statement.set_generic_parameters(generic_parameters);
        }

        if let (
            Some(local),
            Some(function),
            Some(opening_parenthese),
            Some(closing_parenthese),
            Some(end),
        ) = (
            local_token,
            self.function,
            self.opening_parenthese,
            self.closing_parenthese,
            self.end,
        ) {
            statement.set_tokens(LocalFunctionTokens {
                local,
                function_body: FunctionBodyTokens {
                    function,
                    opening_parenthese,
                    closing_parenthese,
                    end,
                    parameter_commas: self.parameter_commas,
                    variable_arguments: self.variable_arguments,
                    variable_arguments_colon: self.variable_arguments_colon,
                    return_type_colon: self.return_type_colon,
                },
            });
        }

        statement
    }

    pub(crate) fn is_variadic(&self) -> bool {
        self.is_variadic
    }

    pub(crate) fn set_return_type_colon(&mut self, token: Token) {
        self.return_type_colon = Some(token);
    }

    pub(crate) fn set_return_type(&mut self, r#type: FunctionReturnType) {
        self.return_type = Some(r#type);
    }

    pub(crate) fn set_variable_arguments_token(&mut self, token: Token) {
        self.variable_arguments = Some(token);
    }

    pub(crate) fn set_variadic(&mut self) {
        self.is_variadic = true;
    }

    pub(crate) fn set_variadic_type(&mut self, r#type: FunctionVariadicType) {
        self.is_variadic = true;
        self.variadic_type = Some(r#type);
    }

    pub(crate) fn set_variable_arguments_colon(&mut self, token: Token) {
        self.variable_arguments_colon = Some(token);
    }

    pub(crate) fn push_parameter(&mut self, typed_identifier: TypedIdentifier) {
        self.parameters.push(typed_identifier);
    }

    pub(crate) fn set_parentheses_tokens(&mut self, open: Token, close: Token) {
        self.opening_parenthese = Some(open);
        self.closing_parenthese = Some(close);
    }

    pub(crate) fn set_parameter_commas(&mut self, commas: Vec<Token>) {
        self.parameter_commas = commas;
    }

    pub(crate) fn set_function_token(&mut self, token: Token) {
        self.function = Some(token);
    }

    pub(crate) fn set_end_token(&mut self, token: Token) {
        self.end = Some(token);
    }

    pub(crate) fn set_generic_parameters(&mut self, generic_parameters: GenericParameters) {
        self.generic_parameters = Some(generic_parameters);
    }
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct FunctionBodyTokens {
    pub function: Token,
    pub opening_parenthese: Token,
    pub closing_parenthese: Token,
    pub end: Token,
    pub parameter_commas: Vec<Token>,
    pub variable_arguments: Option<Token>,

    pub variable_arguments_colon: Option<Token>,
    pub return_type_colon: Option<Token>,
}

impl FunctionBodyTokens {
    super::impl_token_fns!(
        target = [function, opening_parenthese, closing_parenthese, end]
        iter = [
            variable_arguments,
            variable_arguments_colon,
            return_type_colon,
            parameter_commas,
        ]
    );
}
