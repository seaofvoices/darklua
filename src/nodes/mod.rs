//! The collection of nodes used for the Lua abstract syntax tree.

mod arguments;
mod block;
mod expressions;
mod function_body;
mod function_call;
mod identifier;
mod statements;
mod token;
mod typed_identifier;
mod types;
mod variable;

pub use arguments::*;
pub use block::*;
pub use expressions::*;
pub use function_body::*;
pub use function_call::*;
pub use identifier::*;
pub use statements::*;
pub use token::*;
pub use typed_identifier::*;
pub use types::*;
pub use variable::*;

macro_rules! impl_token_fns {
    (
        target = [ $( $field:ident ),* $(,)? ]
        $(iter = [ $( $iter_field:ident),* $(,)? ])?
        $(iter_flatten = [ $( $iter_flatten_field:ident),* $(,)? ])?
    ) => {
        pub fn clear_comments(&mut self) {
            $(
                self.$field.clear_comments();
            )*
            $($(
                    for token in self.$iter_field.iter_mut() {
                        token.clear_comments();
                    }
            )*)?
            $($(
                    for token in self.$iter_flatten_field.iter_mut().flatten() {
                        token.clear_comments();
                    }
            )*)?
        }

        pub fn clear_whitespaces(&mut self) {
            $(
                self.$field.clear_whitespaces();
            )*
            $($(
                for token in self.$iter_field.iter_mut() {
                    token.clear_whitespaces();
                }
            )*)?
            $($(
                for token in self.$iter_flatten_field.iter_mut().flatten() {
                    token.clear_whitespaces();
                }
            )*)?
        }

        pub(crate) fn replace_referenced_tokens(&mut self, code: &str) {
            $(
                self.$field.replace_referenced_tokens(code);
            )*
            $($(
                for token in self.$iter_field.iter_mut() {
                    token.replace_referenced_tokens(code);
                }
            )*)?
            $($(
                for token in self.$iter_flatten_field.iter_mut().flatten() {
                    token.replace_referenced_tokens(code);
                }
            )*)?
        }

        pub(crate) fn shift_token_line(&mut self, amount: isize) {
            $(
                self.$field.shift_token_line(amount);
            )*
            $($(
                for token in self.$iter_field.iter_mut() {
                    token.shift_token_line(amount);
                }
            )*)?
            $($(
                for token in self.$iter_flatten_field.iter_mut().flatten() {
                    token.shift_token_line(amount);
                }
            )*)?
        }

        pub(crate) fn filter_comments(&mut self, filter: impl Fn(&crate::nodes::Trivia) -> bool) {
            $(
                self.$field.filter_comments(&filter);
            )*
            $($(
                for token in self.$iter_field.iter_mut() {
                    token.filter_comments(&filter);
                }
            )*)?
            $($(
                for token in self.$iter_flatten_field.iter_mut().flatten() {
                    token.filter_comments(&filter);
                }
            )*)?
        }
    };

    (
        iter = [$($field:ident),*]
        iter_flatten = [$($flatten_field:ident),*]
    ) => {
        $crate::nodes::impl_token_fns!(
            target = []
            iter = [ $($field,)* ]
            iter_flatten = [ $($flatten_field,)* ]
        );
    };

    (
        iter = [$($field:ident),*]
    ) => {
        $crate::nodes::impl_token_fns!(
            target = []
            iter = [ $($field,)* ]
            iter_flatten = []
        );
    };

    (
        iter_flatten = [$($field:ident),*]
    ) => {
        $crate::nodes::impl_token_fns!(
            target = []
            iter = []
            iter_flatten = [ $($field,)* ]
        );
    };
}

pub(crate) use impl_token_fns;
