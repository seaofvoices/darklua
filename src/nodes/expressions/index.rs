use crate::nodes::{Expression, Prefix, Token};

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct IndexExpressionTokens {
    pub opening_bracket: Token,
    pub closing_bracket: Token,
}

impl IndexExpressionTokens {
    pub fn clear_comments(&mut self) {
        self.opening_bracket.clear_comments();
        self.closing_bracket.clear_comments();
    }

    pub fn clear_whitespaces(&mut self) {
        self.opening_bracket.clear_whitespaces();
        self.closing_bracket.clear_whitespaces();
    }

    pub(crate) fn replace_referenced_tokens(&mut self, code: &str) {
        self.opening_bracket.replace_referenced_tokens(code);
        self.closing_bracket.replace_referenced_tokens(code);
    }

    pub(crate) fn shift_token_line(&mut self, amount: usize) {
        self.opening_bracket.shift_token_line(amount);
        self.closing_bracket.shift_token_line(amount);
    }
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct IndexExpression {
    prefix: Prefix,
    index: Expression,
    tokens: Option<IndexExpressionTokens>,
}

impl IndexExpression {
    pub fn new<P: Into<Prefix>, E: Into<Expression>>(prefix: P, expression: E) -> Self {
        Self {
            prefix: prefix.into(),
            index: expression.into(),
            tokens: None,
        }
    }

    pub fn with_tokens(mut self, tokens: IndexExpressionTokens) -> Self {
        self.tokens = Some(tokens);
        self
    }

    #[inline]
    pub fn set_tokens(&mut self, tokens: IndexExpressionTokens) {
        self.tokens = Some(tokens);
    }

    #[inline]
    pub fn get_tokens(&self) -> Option<&IndexExpressionTokens> {
        self.tokens.as_ref()
    }

    #[inline]
    pub fn get_prefix(&self) -> &Prefix {
        &self.prefix
    }

    #[inline]
    pub fn get_index(&self) -> &Expression {
        &self.index
    }

    #[inline]
    pub fn mutate_prefix(&mut self) -> &mut Prefix {
        &mut self.prefix
    }

    #[inline]
    pub fn mutate_index(&mut self) -> &mut Expression {
        &mut self.index
    }

    pub fn clear_comments(&mut self) {
        if let Some(tokens) = &mut self.tokens {
            tokens.clear_comments();
        }
    }

    pub fn clear_whitespaces(&mut self) {
        if let Some(tokens) = &mut self.tokens {
            tokens.clear_whitespaces();
        }
    }

    pub(crate) fn replace_referenced_tokens(&mut self, code: &str) {
        if let Some(tokens) = &mut self.tokens {
            tokens.replace_referenced_tokens(code);
        }
    }

    pub(crate) fn shift_token_line(&mut self, amount: usize) {
        if let Some(tokens) = &mut self.tokens {
            tokens.shift_token_line(amount);
        }
    }
}
