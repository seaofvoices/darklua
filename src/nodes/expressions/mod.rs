mod binary;
mod field;
mod function;
mod if_expression;
mod index;
mod interpolated_string;
mod number;
mod parenthese;
mod prefix;
mod string;
pub(crate) mod string_utils;
mod table;
mod type_cast;
mod unary;

pub use binary::*;
pub use field::*;
pub use function::*;
pub use if_expression::*;
pub use index::*;
pub use interpolated_string::*;
pub use number::*;
pub use parenthese::*;
pub use prefix::*;
pub use string::*;
pub use string_utils::StringError;
pub use table::*;
pub use type_cast::*;
pub use unary::*;

use crate::nodes::{FunctionCall, Identifier, Token, Variable};

use super::impl_token_fns;

use std::num::FpCategory;

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum Expression {
    Binary(Box<BinaryExpression>),
    Call(Box<FunctionCall>),
    False(Option<Token>),
    Field(Box<FieldExpression>),
    Function(FunctionExpression),
    Identifier(Identifier),
    If(Box<IfExpression>),
    Index(Box<IndexExpression>),
    Nil(Option<Token>),
    Number(NumberExpression),
    Parenthese(Box<ParentheseExpression>),
    String(StringExpression),
    InterpolatedString(InterpolatedStringExpression),
    Table(TableExpression),
    True(Option<Token>),
    Unary(Box<UnaryExpression>),
    VariableArguments(Option<Token>),
    TypeCast(TypeCastExpression),
}

impl Expression {
    #[inline]
    pub fn nil() -> Self {
        Self::Nil(None)
    }

    #[inline]
    pub fn variable_arguments() -> Self {
        Self::VariableArguments(None)
    }

    pub fn identifier<S: Into<Identifier>>(identifier: S) -> Self {
        Self::Identifier(identifier.into())
    }

    pub fn in_parentheses(self) -> Self {
        Self::Parenthese(ParentheseExpression::new(self).into())
    }
}

impl From<bool> for Expression {
    fn from(boolean: bool) -> Expression {
        if boolean {
            Expression::True(None)
        } else {
            Expression::False(None)
        }
    }
}

impl From<f64> for Expression {
    fn from(value: f64) -> Expression {
        match value.classify() {
            FpCategory::Nan => BinaryExpression::new(
                BinaryOperator::Slash,
                DecimalNumber::new(0.0),
                DecimalNumber::new(0.0),
            )
            .into(),
            FpCategory::Infinite => BinaryExpression::new(
                BinaryOperator::Slash,
                Expression::from(if value.is_sign_positive() { 1.0 } else { -1.0 }),
                DecimalNumber::new(0.0),
            )
            .into(),
            FpCategory::Zero => {
                DecimalNumber::new(if value.is_sign_positive() { 0.0 } else { -0.0 }).into()
            }
            FpCategory::Subnormal | FpCategory::Normal => {
                if value < 0.0 {
                    UnaryExpression::new(UnaryOperator::Minus, Expression::from(value.abs())).into()
                } else if value < 0.1 {
                    let exponent = value.log10().floor();
                    let new_value = value / 10_f64.powf(exponent);

                    DecimalNumber::new(new_value)
                        .with_exponent(exponent as i64, true)
                        .into()
                } else if value > 999.0 && (value / 100.0).fract() == 0.0 {
                    let mut exponent = value.log10().floor();
                    let mut power = 10_f64.powf(exponent);

                    while exponent > 2.0 && (value / power).fract() != 0.0 {
                        exponent -= 1.0;
                        power /= 10.0;
                    }

                    DecimalNumber::new(value / power)
                        .with_exponent(exponent as i64, true)
                        .into()
                } else {
                    DecimalNumber::new(value).into()
                }
            }
        }
    }
}

impl From<f32> for Expression {
    fn from(value: f32) -> Self {
        (value as f64).into()
    }
}

impl From<usize> for Expression {
    fn from(value: usize) -> Self {
        (value as f64).into()
    }
}

impl From<u64> for Expression {
    fn from(value: u64) -> Self {
        (value as f64).into()
    }
}

impl From<u32> for Expression {
    fn from(value: u32) -> Self {
        (value as f64).into()
    }
}

impl From<u16> for Expression {
    fn from(value: u16) -> Self {
        (value as f64).into()
    }
}

impl From<u8> for Expression {
    fn from(value: u8) -> Self {
        (value as f64).into()
    }
}

impl From<i64> for Expression {
    fn from(value: i64) -> Self {
        (value as f64).into()
    }
}

impl From<i32> for Expression {
    fn from(value: i32) -> Self {
        (value as f64).into()
    }
}

impl From<i16> for Expression {
    fn from(value: i16) -> Self {
        (value as f64).into()
    }
}

impl From<i8> for Expression {
    fn from(value: i8) -> Self {
        (value as f64).into()
    }
}

impl From<BinaryExpression> for Expression {
    fn from(binary: BinaryExpression) -> Expression {
        Expression::Binary(Box::new(binary))
    }
}

impl From<FunctionCall> for Expression {
    fn from(call: FunctionCall) -> Expression {
        Expression::Call(Box::new(call))
    }
}

impl From<FieldExpression> for Expression {
    fn from(field: FieldExpression) -> Expression {
        Expression::Field(Box::new(field))
    }
}

impl From<FunctionExpression> for Expression {
    fn from(function: FunctionExpression) -> Self {
        Expression::Function(function)
    }
}

impl From<Identifier> for Expression {
    fn from(identifier: Identifier) -> Self {
        Expression::Identifier(identifier)
    }
}

impl From<IfExpression> for Expression {
    fn from(if_expression: IfExpression) -> Expression {
        Expression::If(Box::new(if_expression))
    }
}

impl From<IndexExpression> for Expression {
    fn from(index: IndexExpression) -> Self {
        Self::Index(Box::new(index))
    }
}

impl From<NumberExpression> for Expression {
    fn from(number: NumberExpression) -> Self {
        Self::Number(number)
    }
}

impl From<DecimalNumber> for Expression {
    fn from(number: DecimalNumber) -> Self {
        Self::Number(NumberExpression::Decimal(number))
    }
}

impl From<HexNumber> for Expression {
    fn from(number: HexNumber) -> Self {
        Self::Number(NumberExpression::Hex(number))
    }
}

impl From<BinaryNumber> for Expression {
    fn from(number: BinaryNumber) -> Self {
        Self::Number(NumberExpression::Binary(number))
    }
}

impl From<Prefix> for Expression {
    fn from(prefix: Prefix) -> Self {
        match prefix {
            Prefix::Call(call) => Self::Call(Box::new(call)),
            Prefix::Field(field) => Self::Field(field),
            Prefix::Identifier(name) => Self::Identifier(name),
            Prefix::Index(index) => Self::Index(index),
            Prefix::Parenthese(expression) => expression.into(),
        }
    }
}

impl From<ParentheseExpression> for Expression {
    fn from(expression: ParentheseExpression) -> Self {
        Self::Parenthese(expression.into())
    }
}

impl From<StringExpression> for Expression {
    fn from(string: StringExpression) -> Self {
        Self::String(string)
    }
}

impl From<InterpolatedStringExpression> for Expression {
    fn from(interpolated_string: InterpolatedStringExpression) -> Self {
        Self::InterpolatedString(interpolated_string)
    }
}

impl From<TableExpression> for Expression {
    fn from(table: TableExpression) -> Self {
        Self::Table(table)
    }
}

impl From<UnaryExpression> for Expression {
    fn from(unary: UnaryExpression) -> Self {
        Self::Unary(Box::new(unary))
    }
}

impl From<TypeCastExpression> for Expression {
    fn from(type_cast: TypeCastExpression) -> Self {
        Self::TypeCast(type_cast)
    }
}

impl From<Variable> for Expression {
    fn from(variable: Variable) -> Self {
        match variable {
            Variable::Identifier(identifier) => Self::Identifier(identifier),
            Variable::Field(field) => Self::Field(field),
            Variable::Index(index) => Self::Index(index),
        }
    }
}

impl<T: Into<Expression>> From<Option<T>> for Expression {
    fn from(value: Option<T>) -> Self {
        match value {
            None => Self::nil(),
            Some(value) => value.into(),
        }
    }
}

#[cfg(test)]
mod test {
    macro_rules! snapshot_from_expression {
        ($($name:ident => $input:expr),+ $(,)?) => {
            $(
                #[test]
                fn $name() {
                    let result = crate::nodes::Expression::from($input);

                    insta::assert_debug_snapshot!(stringify!($name), result);
                }
            )+
        };
    }

    mod expression_from_floats {
        snapshot_from_expression!(
            f64_0 => 0_f64,
            f64_1e42 => 1e42_f64,
            f64_infinity => f64::INFINITY,
            i64_minus_one => -1_i64,
            f64_minus_zero => -0.0,
        );
    }
}
