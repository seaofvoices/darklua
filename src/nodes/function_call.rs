use crate::nodes::{Arguments, Expression, Identifier, Prefix, Token};

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct FunctionCallTokens {
    pub colon: Option<Token>,
}

impl FunctionCallTokens {
    super::impl_token_fns!(iter = [colon]);
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct FunctionCall {
    prefix: Box<Prefix>,
    arguments: Arguments,
    method: Option<Identifier>,
    tokens: Option<FunctionCallTokens>,
}

impl FunctionCall {
    pub fn new(prefix: Prefix, arguments: Arguments, method: Option<Identifier>) -> Self {
        Self {
            prefix: Box::new(prefix),
            arguments,
            method,
            tokens: None,
        }
    }

    pub fn from_name<T: Into<Identifier>>(name: T) -> Self {
        Self {
            prefix: Box::new(name.into().into()),
            arguments: Arguments::default(),
            method: None,
            tokens: None,
        }
    }

    pub fn from_prefix<T: Into<Prefix>>(prefix: T) -> Self {
        Self {
            prefix: Box::new(prefix.into()),
            arguments: Arguments::default(),
            method: None,
            tokens: None,
        }
    }

    pub fn with_tokens(mut self, tokens: FunctionCallTokens) -> Self {
        self.tokens = Some(tokens);
        self
    }

    #[inline]
    pub fn set_tokens(&mut self, tokens: FunctionCallTokens) {
        self.tokens = Some(tokens);
    }

    #[inline]
    pub fn get_tokens(&self) -> Option<&FunctionCallTokens> {
        self.tokens.as_ref()
    }

    pub fn with_arguments<A: Into<Arguments>>(mut self, arguments: A) -> Self {
        self.arguments = arguments.into();
        self
    }

    pub fn with_argument<T: Into<Expression>>(mut self, argument: T) -> Self {
        self.arguments = self.arguments.with_argument(argument);
        self
    }

    pub fn with_method<IntoString: Into<Identifier>>(mut self, method: IntoString) -> Self {
        self.method.replace(method.into());
        self
    }

    #[inline]
    pub fn get_arguments(&self) -> &Arguments {
        &self.arguments
    }

    #[inline]
    pub fn get_method(&self) -> Option<&Identifier> {
        self.method.as_ref()
    }

    #[inline]
    pub fn get_prefix(&self) -> &Prefix {
        &self.prefix
    }

    #[inline]
    pub fn take_method(&mut self) -> Option<Identifier> {
        self.method.take()
    }

    #[inline]
    pub fn set_arguments(&mut self, arguments: Arguments) {
        self.arguments = arguments;
    }

    #[inline]
    pub fn set_method(&mut self, method: Identifier) {
        self.method.replace(method);
    }

    #[inline]
    pub fn mutate_arguments(&mut self) -> &mut Arguments {
        &mut self.arguments
    }

    #[inline]
    pub fn mutate_prefix(&mut self) -> &mut Prefix {
        &mut self.prefix
    }

    super::impl_token_fns!(iter = [tokens, method]);
}
