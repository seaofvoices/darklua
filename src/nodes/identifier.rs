use crate::nodes::Token;

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Identifier {
    name: String,
    token: Option<Token>,
}

impl Identifier {
    pub fn new<S: Into<String>>(name: S) -> Self {
        Self {
            name: name.into(),
            token: None,
        }
    }

    pub fn with_token(mut self, token: Token) -> Self {
        self.token = Some(token);
        self
    }

    #[inline]
    pub fn set_token(&mut self, token: Token) {
        self.token = Some(token);
    }

    #[inline]
    pub fn get_token(&self) -> Option<&Token> {
        self.token.as_ref()
    }

    #[inline]
    pub fn get_name(&self) -> &String {
        &self.name
    }

    #[inline]
    pub fn mutate_name(&mut self) -> &mut String {
        &mut self.name
    }

    #[inline]
    pub fn set_name<IntoString: Into<String>>(&mut self, name: IntoString) {
        let name = name.into();
        if let Some(token) = &mut self.token {
            token.replace_with_content(name.clone());
        }
        self.name = name;
    }

    #[inline]
    pub fn into_name(self) -> String {
        self.name
    }

    pub fn clear_comments(&mut self) {
        if let Some(token) = &mut self.token {
            token.clear_comments();
        }
    }

    pub fn clear_whitespaces(&mut self) {
        if let Some(token) = &mut self.token {
            token.clear_whitespaces();
        }
    }

    pub(crate) fn replace_referenced_tokens(&mut self, code: &str) {
        if let Some(token) = &mut self.token {
            token.replace_referenced_tokens(code);
        }
    }

    pub(crate) fn shift_token_line(&mut self, amount: usize) {
        if let Some(token) = &mut self.token {
            token.shift_token_line(amount);
        }
    }
}

impl<IntoString: Into<String>> From<IntoString> for Identifier {
    fn from(identifier: IntoString) -> Self {
        Self {
            name: identifier.into(),
            token: None,
        }
    }
}

#[cfg(test)]
mod test {
    use crate::nodes::Position;

    use super::*;

    #[test]
    fn set_name_replaces_the_token_content() {
        let token = Token::new_with_line(7, 10, 1);
        let mut identifier = Identifier::new("var").with_token(token);

        identifier.set_name("newVar");

        assert_eq!(
            identifier.get_token().unwrap(),
            &Token::from_position(Position::LineNumber {
                line_number: 1,
                content: "newVar".into(),
            })
        );
    }
}
