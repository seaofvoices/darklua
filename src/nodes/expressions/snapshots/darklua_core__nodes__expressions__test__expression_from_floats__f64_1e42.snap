---
source: src/nodes/expressions/mod.rs
expression: result
---
Number(
    Decimal(
        DecimalNumber {
            float: 1.0,
            exponent: Some(
                (
                    42,
                    true,
                ),
            ),
            token: None,
        },
    ),
)
